drivers/ethernet/dwc_xgmac/eth_dwc_xgmac.c

```diff
@@ -0,0 +1,1929 @@
+	mem_addr_t ioaddr = get_reg_base_addr(dev);
+	mem_addr_t reg_addr;
+	uint32_t reg_val;
+
+	if (!data->dev_started || data->link_speed == LINK_DOWN ||
+	    (!net_if_flag_is_set(data->iface, NET_IF_UP))) {
+		dma_int_status =
+			sys_read32(ioaddr + XGMAC_DMA_BASE_ADDR_OFFSET + DMA_INTERRUPT_STATUS_OFST);
+		for (uint32_t x = 0; x < config->num_dma_chnl; x++) {
+			if (dma_int_status & BIT(x)) {
+				LOG_ERR("%s ignoring dma ch %d interrupt: %x ", dev->name, x,
+					sys_read32(ioaddr + XGMAC_DMA_CHNLx_BASE_ADDR_OFFSET(x) +
+						   DMA_CHx_STATUS_OFST));
+				reg_val = DMA_CHx_STATUS_NIS_SET_MSK | DMA_CHx_STATUS_AIS_SET_MSK |
+					  DMA_CHx_STATUS_CDE_SET_MSK | DMA_CHx_STATUS_FBE_SET_MSK |
+					  DMA_CHx_STATUS_DDE_SET_MSK | DMA_CHx_STATUS_RPS_SET_MSK |
+					  DMA_CHx_STATUS_RBU_SET_MSK | DMA_CHx_STATUS_TBU_SET_MSK |
+					  DMA_CHx_STATUS_TPS_SET_MSK | DMA_CHx_STATUS_RI_SET_MSK |
+					  DMA_CHx_STATUS_TI_SET_MSK;
+				sys_write32(reg_val, (ioaddr + XGMAC_DMA_CHNLx_BASE_ADDR_OFFSET(x) +
+						      DMA_CHx_STATUS_OFST));
+			}
+		}
+
+		LOG_ERR("%s ignoring xgmac interrupt: device not started,link is down or network "
+			"interface is not up",
+			dev->name);
+
+		return;
+	}
+
+	/* Interrupt Top half processing
+	 */
+	reg_addr = (ioaddr + XGMAC_DMA_BASE_ADDR_OFFSET + DMA_INTERRUPT_STATUS_OFST);
+	/* Only set the interrupt, do not overwrite the interrupt status stored in the context.
+	 * The status will be cleared once the corresponding action is completed in the work item
+	 */
+	cntxt_data->dma_interrupt_sts |= sys_read32(reg_addr);
+	for (uint32_t x = 0; x < config->num_dma_chnl; x++) {
+		if (cntxt_data->dma_interrupt_sts & BIT(x)) {
+			reg_addr = (ioaddr + XGMAC_DMA_CHNLx_BASE_ADDR_OFFSET(x) +
+				    DMA_CHx_STATUS_OFST);
+			dmach_interrupt_sts = sys_read32(reg_addr);
+			sys_write32(dmach_interrupt_sts, reg_addr);
+#ifdef CONFIG_ETH_DWC_XGMAC_BOTTOM_HALF_WORK_QUEUE
+			/* Only set the interrupt, do not overwrite the interrupt status stored in
+			 * the context. The status will be cleared once the corresponding action is
+			 * done in the work item
+			 */
+			cntxt_data->dma_chnl_interrupt_sts[x] |= dmach_interrupt_sts;
+#else
+			eth_dwc_xgmac_dmach_isr(dev, dmach_interrupt_sts, x);
+			WRITE_BIT(cntxt_data->dma_interrupt_sts, x, 0);
+#endif
+		}
+	}
+
+	reg_addr = (ioaddr + XGMAC_MTL_BASE_ADDR_OFFSET + MTL_INTERRUPT_STATUS_OFST);
+	reg_val = sys_read32(reg_addr);
+#ifdef CONFIG_ETH_DWC_XGMAC_BOTTOM_HALF_WORK_QUEUE
+	cntxt_data->mtl_interrupt_sts |= reg_val;
+#else
+	eth_dwc_xgmac_mtl_isr(dev, reg_val);
+#endif
+
+	reg_addr = (ioaddr + XGMAC_CORE_BASE_ADDR_OFFSET + CORE_MAC_INTERRUPT_STATUS_OFST);
+	reg_val = sys_read32(reg_addr);
+#ifdef CONFIG_ETH_DWC_XGMAC_BOTTOM_HALF_WORK_QUEUE
+	cntxt_data->mac_interrupt_sts |= reg_val;
+#else
+	eth_dwc_xgmac_mac_isr(dev, reg_val);
+#endif
+
+#ifdef CONFIG_ETH_DWC_XGMAC_BOTTOM_HALF_WORK_QUEUE
+	/*
+	 * submitting work item to work queue for interrupt bottom half processing.
```

Comment:
this comment not required.

Reply:
the driver is developed to support both with and without bottom-half support. user has the freedom to choose without the bottom as well.